src/1-theory/10-async.adoc

```diff
@@ -0,0 +1,107 @@
+they are all following a request and response structure.
+
+Getting a response from a request is one way to do things, but there are lots of
+situations where a client might expect more than one response, or a response with
+more information later as more work happens on the server side.
+
+In the example of a HTTP API where a user is buying something from your site. They make
+a payment, and they want to know if it has been accepted or not.
+
+.A completely synchronous approach to handling a payment from a GUI, with your API accepting the information and sending it off to an external API acting as a payment gateway.
+[plantuml]
+---------------------------------------------------------------------
+@startuml
+
+actor User
+boundary "Web GUI" as GUI
+control "API"
+boundary "Payment Gateway" as PG
+
+User -> API: POST /payments
+API -> PG: POST /submitPaymentAttempt
+PG -> PG: May take 30-60s to confirm
+API <- PG: HTTP 200 OK
+GUI <- API: Web GUI gets the confirmation
+User <- GUI: User gets confirmation
+
+@enduml
+---------------------------------------------------------------------
+
+If the world was a perfect place, this would be fine. All messages would somehow move faster than the speed of light, all servers would alway be responding perfectly, and no messages would be lost in transit, meaning the user gets a nice, quick, consistent experience and the user interface does not leave them guessing.
+
+== Job Queues
+
+Sadly none of this is true, but people seem to  design their data flows like it might be. The above flow will leave the user sat there twiddling their thumbs for however long it takes that API gateway to respond.
+
+There is one issue here that the user is twiddling their thumbs for 30 seconds, but a bigger issue is that the API server is twiddling its thumbs for 30 seconds. Your API only has so many threads available to do stuff, so if you are wasting one of them sitting around for 30-60 seconds, that is a lot of other requests not being handled. More on that later.
+
+One way to avoid the user and/or API server twiddling literal or figurative thumbs, is to have the API immediately pass that work off to a a job queue like RabbitMQ, Sidekiq, etc. This lets the API respond with a "we are working on it" response, then have the server respond with a "Yeah yeah we're working on it" response.
+
+In HTTP this is done with a `201 Created`, or `202 Accepted` response code, which lets the client know they should look for some sort of link to get their updates.
+
+.Letting the API do a bit less work up front, shiting the waiting work to a job queue, and letting the GUI check for updates.
+[plantuml]
+----
+@startuml
+
+title Message Style - Sequence Diagram
+
+actor User
+boundary "Web GUI" as GUI
+control "API"
+control Worker
+boundary "Payment Gateway" as PG
+
+User -> API: POST /payments
+API -> Worker: Queue Job
+GUI <- API: HTTP 201 Accepted (with a URL to poll)
```

Comment:
I think you want `Created` here or change it to `202`. :)

Reply:
Get the fuck out of here bryan, dropping by with an annoyingly correct #WellActually.

Reply:
My bad! 🤓